### PR TITLE
Fixed NonZero to have sorted result.

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
 - Fixed faulty GPU nonzero kernel behavior : Previously it was generating result indices in a random order, thought it should be sorted
 - Fixed NonZero cldnn unittest to compare with ngraph reference code.

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/gather_nonzero_ref.cl [hunk 1]
```diff
@@ -5,83 +5,150 @@
...
+#else
+    __global OUTPUT_TYPE* out_mem = output;
+#endif
+
+    int b_inc, f_inc, y_inc, x_inc;
+    int count_nzero = output_shape[1];
+#if OV_INPUT_RANK == 1 // b
+    #define ADD_IDXS \
+        int b = input_idx_v; \
+        int b = input_idx_v / INPUT0_BATCH_PITCH; \
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/gather_nonzero_ref.cl:32]: It seems like that these two lines raise a redefinition error.

File: src/plugins/intel_gpu/src/kernel_selector/core/kernel_selector_params.h [hunk 2]
```diff
@@ -580,7 +580,7 @@ struct base_params : public Params {
     std::vector<fused_operation_desc> fused_ops = {};
     MultiDataTensor inputs;
     MultiDataTensor outputs;
-
+    int32_t ov_input_rank; // original input rank
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/kernel_selector/core/kernel_selector_params.h:583]: Why do we need that in the base class? Let's move to non zero params only..

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h [hunk 3]
```diff
@@ -141,12 +142,16 @@ struct kernel_impl_params {
...
-                       , input_layouts(_int_layouts)
+                       , input_layouts(_in_layouts)
                        , output_layout(_out_layout)
                        , fused_desc(_fused_descs)
                        , fused_act_funcs(_fused_act_funcs)
                        , activation_params(_act_params)
-                       , primary_input_idx(0) {}
+                       , primary_input_idx(0) {
+        if (!_in_layouts.empty() && !_in_layouts[0].get_partial_shape().rank().is_dynamic()) {
+            ov_input_rank = _in_layouts[0].get_shape().size();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h:152]: Again, why do we need that as common field?
To me it looks like all you need is to modify gather_nonzero_impl::create and count_nonzero_impl::create functions and kernel parameters for those 2 operations to pass original rank that may be taken in the moment of conversion from impl_params.input_layouts.

**Review by vladimir-paramuzov — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h (hunk 3 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/include/kernel_selector_helper.h]: I agree, I fixed. Thanks.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/gather_nonzero_ref.cl (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/gather_nonzero_ref.cl]: Thnks, I fixed and added 1d,2d,3d tests too

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/non_zero/gather_nonzero_kernel_ref.cpp [hunk 4]
```diff
@@ -44,37 +44,15 @@ KernelsData GatherNonzeroKernelRef::GetKernelsData(const Params& params, const o
 
     auto entry_point = GetEntryPoint(kernelName, newParams.layerID, params, options);
     auto cldnn_jit = MakeBaseParamsJitConstants(newParams);
+    cldnn_jit.AddConstant(MakeJitConstant("OV_INPUT_RANK", newParams.ov_input_rank));
+    cldnn_jit.AddConstant(MakeJitConstant("TOTAL_DATA_SIZE", newParams.inputs[0].LogicalSize()));
+    if (newParams.inputs[0].LogicalSize() * newParams.ov_input_rank < (64*1024)/(newParams.outputs[0].ElementSize()))
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/non_zero/gather_nonzero_kernel_ref.cpp:49]: please replace magic constant by variable with meaningful name.

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/non_zero/gather_nonzero_kernel_ref.cpp:49]: And define it in kernel via jitter too

Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/non_zero/gather_nonzero_kernel_ref.cpp]: Fixed it and make it to be read from param.

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**